boutdata/collect.py

```diff
@@ -307,392 +239,823 @@ def getDataFile(i):
+        global_slices.append(0)
+    if "x" in dimensions:
+        global_slices.append(slice(xgstart, xgstop))
+    elif parallel_read:
+        global_slices.append(0)
+    if "y" in dimensions:
+        global_slices.append(slice(ygstart, ygstop))
+    elif parallel_read:
+        global_slices.append(0)
+    if "z" in dimensions:
+        global_slices.append(slice(None))
+    elif parallel_read:
+        global_slices.append(0)
+    global_slices = tuple(global_slices)
 
-    if datafile_cache is None:
-        # close the DataFile if we are not keeping it in a cache
-        f.close()
+    if info:
+        sys.stdout.write(
+            "\rReading from {}: [{}-{}][{}-{}] -> [{}-{}][{}-{}]\n".format(
+                i,
+                xstart,
+                xstop - 1,
+                ystart,
+                ystop - 1,
+                xgstart,
+                xgstop - 1,
+                ygstart,
+                ygstop - 1
+            )
+        )
+
+    if is_fieldperp:
+        f_attributes = datafile.attributes(varname)
+        temp_yindex = f_attributes["yindex_global"]
+        if temp_yindex < 0:
+            # No data for FieldPerp on this processor
+            return None, None
+
+    result[global_slices] = datafile.read(varname, ranges=local_slices)
+
+    if is_fieldperp:
+        return temp_yindex, f_attributes
+
+    return None, None
+
+
+def _get_x_range(xguards, xind, pe_xind, nxpe, mxsub, mxg, inrange):
+    """
+    Get local ranges of x-indices
 
-    # Map between dimension names and output size
-    sizes = {'x': xsize, 'y': ysize, 'z': zsize, 't': tsize}
+    Parameters
+    ----------
+    xguards : bool
+        Include x-boundaries?
+    xind : slice
+        Global slice to apply to x-dimension
+    pe_xind : int
+        x-index of the processor
+    nxpe : int
+        Number of processors in the x-direction
+    mxsub : int
+        Number of grid cells (excluding guard cells) in the x-direction on a single
+        procssor
+    mxg : int
+        Number of guard cells in the x-direction
+    inrange : bool
+        Does the processor have data to read?
 
-    # Create a list with size of each dimension
-    ddims = [sizes[d] for d in dimensions]
+    Returns
+    -------
+    xstart : int
+        Local x-index to start reading
+    xstop : int
+        Local x-index to stop reading
+    xgstart : int
+        Global x-index to start putting data
+    xgstop : int
+        Global x-index to stop putting data
+    inrange : bool
+        Updated version of inrange - changed to False if this processor has no data to
+        read
+    """
+    # Local ranges
+    if xguards:
+        xstart = xind.start - pe_xind * mxsub
+        xstop = xind.stop - pe_xind * mxsub
+
+        # Check lower x boundary
+        if pe_xind == 0:
+            # Keeping inner boundary
+            if xstop <= 0:
+                inrange = False
+            if xstart < 0:
+                xstart = 0
```

Comment:
This little bit is repeated several times with variations, my refactoring fingers are itchy to try and reduce the duplication, but I can't see a way that doesn't just make it worse.

Possibly to use `xstart = max(0, xstart)` instead of the second conditional? That's probably less readable than the current version though

Reply:
I thought of a way to refactor out the checks into two functions that can be used by both `_get_x_range()` and `_get_y_range()`. I think refactoring is a good idea, because in doing so I fixed a bug with y-slicing that had been hiding another bug (updating `inrange` with `_get_y_range()` for `FieldPerp`s, which should not be affected by `yind`) from the tests! Will push commits in a few mins when I check all the other tests pass.